File: 02-week-2--recursion/05-day-5--recursive-selection-sort/benchmark/benchmark.rb
# recursive solution
def selection_sort_recursive(arr)
  return [] if arr.length == 0

  min = arr.min
  idx = arr.index(min)
  arr.delete_at(idx)

  sorted = selection_sort_recursive(arr)
  sorted.unshift(min)
end

# iterative solution
def selection_sort(arr)
  sorted = []
  # iterate through the entire array to find the minimum element
  # push the minimum element onto the new sorted array
  # delete the element at the minimum element index 
  # repeat until the array is empty
  until arr.length == 0
    min = arr.min
    idx = arr.index(min)
    sorted << min
    arr.delete_at(idx)
  end
  sorted
end

def benchmark
  start_time = Time.now

  1000.times do 
    yield
  end

  (Time.now - start_time) / 1000
end

puts "Iterative Solution: #{benchmark { selection_sort([2, 3, 4, 1, 4, 56, -2, 20]) }}"
puts "Recursive Solution: #{benchmark { selection_sort_recursive([2, 3, 4, 1, 4, 56, -2, 20]) }}"

File: 02-week-2--recursion/05-day-5--recursive-selection-sort/benchmark/benchmark.js
// iterative solution
function selectionSort(arr) {
  const sorted = [];
  while (arr.length > 0) {
    const min = Math.min(...arr);
    const idx = arr.indexOf(min);

    sorted.push(min);
    arr.splice(idx, 1);
  }
  return sorted;
}

// recursive solution
function selectionSortRecursive(arr) {
  if (arr.length === 0) {
    return [];
  }
  const min = Math.min(...arr);
  const idx = arr.indexOf(min);
  arr.splice(idx, 1);
  const result = selectionSortRecursive(arr);
  result.unshift(min);
  return result;
}

function benchmark(callback) {
  const startTime = Date.now();

  for (let i = 0; i < 1000; ++i) {
    callback();
  }

  return (Date.now() - startTime) / 1000;
}

console.log('Iterative Solution:', benchmark(() => selectionSort([2, 3, 4, 1, 4, 56, -2, 20])));
console.log('Recursive Solution:', benchmark(() => selectionSortRecursive([2, 3, 4, 1, 4, 56, -2, 20])));
